### PR TITLE
add an option to slice long titles

### PR DESCRIPTION
**Pull request by dyphire**
~~If specialization is required, `forced` and `external` should also be dealt. Don't just deal `default`, that's really a bit ugly.~~

### CODE REVIEW

**Comment by butterw** [comment]
There is a typo in audio bitrate, should be kbps not `kpbs`:
if track['demux-bitrate'] then h(string.format('%.3g kpbs', track['demux-bitrate'] / 1000)) end

**Comment by dyphire** [comment]
> There is a typo in audio bitrate, should be kbps not `kpbs`: if track['demux-bitrate'] then h(string.format('%.3g kpbs', track['demux-bitrate'] / 1000)) end

Fixed.

File: lua/dyn_menu.lua [hunk 1]
```diff
@@ -13,9 +13,16 @@
...
 
+local opts = require("mp.options")
 local menu_prop = 'user-data/menu/items'
 local menu_items = mp.get_property_native(menu_prop, {})
 
+local o = {
+    --slice long titles with specified string length
+    slice_longtitles_amount = 80,
+}
+opts.read_options(o)
```
Comment (tsl0922) [inline lua/dyn_menu.lua:24]:
```suggestion
opts.read_options(o, 'menu')
```

File: lua/dyn_menu.lua [hunk 2]
```diff
@@ -40,6 +47,9 @@ function build_track_title(track, prefix, filename)
         if filename ~= '' then title = title:gsub(filename .. '%.?', '') end
         if title:lower() == codec:lower() then title = '' end
     end
+    if track.title and track.title:len() > o.slice_longtitles_amount + 5 then
```
Comment (tsl0922) [inline lua/dyn_menu.lua:50]: It should be `title` here?

File: lua/dyn_menu.lua [hunk 3]
```diff
@@ -54,11 +64,11 @@ function build_track_title(track, prefix, filename)
     if track['demux-fps'] then h(string.format('%.5g fps', track['demux-fps'])) end
     if track['audio-channels'] then h(track['audio-channels'] .. ' ch') end
     if track['demux-samplerate'] then h(string.format('%.3g kHz', track['demux-samplerate'] / 1000)) end
-    if track['demux-bitrate'] then h(string.format('%.3g kpbs', track['demux-bitrate'] / 1000)) end
+    if track['demux-bitrate'] then h(string.format('%.3g kbps', track['demux-bitrate'] / 1000)) end
     if track.forced then h('forced') end
+    if track.default then h('default') end
```
Comment (tsl0922) [inline lua/dyn_menu.lua:69]: Unrelated change

File: lua/dyn_menu.lua [hunk 4]
```diff
@@ -54,11 +64,11 @@ function build_track_title(track, prefix, filename)
...
     if track.external then h('external') end
     if #hints > 0 then title = string.format('%s [%s]', title, table.concat(hints, ', ')) end
-    if track.default then title = title .. ' (*)' end
```
Comment (tsl0922) [inline lua/dyn_menu.lua:61]: Unrelated change

File: lua/dyn_menu.lua [hunk 5]
```diff
@@ -139,6 +149,9 @@ function update_chapters_menu(submenu)
         local pos = mp.get_property_number('chapter', -1)
         for id, chapter in ipairs(chapter_list) do
             local title = chapter.title or ''
+            if title and title:len() > o.slice_longtitles_amount + 5 then
```
Comment (tsl0922) [inline lua/dyn_menu.lua:152]:
```suggestion
            if title:len() > o.slice_longtitles_amount + 5 then
```

File: lua/dyn_menu.lua (hunk 2 above)
Comment (dyphire) [reply lua/dyn_menu.lua]: Sure

File: lua/dyn_menu.lua [hunk 6]
```diff
@@ -40,6 +47,9 @@ function build_track_title(track, prefix, filename)
         if filename ~= '' then title = title:gsub(filename .. '%.?', '') end
         if title:lower() == codec:lower() then title = '' end
     end
+    if title and title:len() > o.slice_longtitles_amount + 5 then
```
Comment (tsl0922) [inline lua/dyn_menu.lua:50]:
```suggestion
    if title:len() > o.slice_longtitles_amount + 5 then
```

File: lua/dyn_menu.lua (hunk 3 above)
Comment (dyphire) [reply lua/dyn_menu.lua:69]: Split into a separate commit.

Comment (tsl0922) [reply lua/dyn_menu.lua:69]: Why do you want to revert it? I do think `(*)` is more readable than `default`

Comment (dyphire) [reply lua/dyn_menu.lua:69]: Although the `default` property are also printed in this way of `(*)` in the console, these properties of `forced` and `external` are not displayed in the console.
When rendering in the menu, that is not coordinated to only specialize the default property because we display all of them simultaneously.
If you have a way to handle `forced` and `external` as well, I don't object to retaining the current behavior.

Comment (dyphire) [reply lua/dyn_menu.lua]: Remove the commit, there is a issue with the idea here.

File: lua/dyn_menu.lua (hunk 1 above)
Comment (dyphire) [reply lua/dyn_menu.lua:24]: Incorrect suggestions can lead to the following issues:
```log
[w][dyn_menu] script-opts/menu.conf:3 unknown key 'uosc', ignoring 
```

Comment (tsl0922) [reply lua/dyn_menu.lua:24]: Then you will need to configure it with `dyn_menu.conf`. I want to use the same file name for both the DLL and lua script.

To remove this warning, you may add the `uosc` option here with a comment saying it is used by the cplugin.

File: lua/dyn_menu.lua (hunk 3 above)
Comment (tsl0922) [reply lua/dyn_menu.lua]: You may want to check mpv's print_stream function:

https://github.com/mpv-player/mpv/blob/41259db9523bede8c656c7a8664484e0e4befa02/player/loadfile.c#L241-L298

File: lua/dyn_menu.lua (hunk 1 above)
Comment (dyphire) [reply lua/dyn_menu.lua:24]: I don't like this solution, I prefer to keep different configuration files.

File: lua/dyn_menu.lua (hunk 3 above)
Comment (dyphire) [reply lua/dyn_menu.lua]: That's exactly why I added a new commit. The issue here is that the meaning of abbreviations is unclear, and more than one person on the mpv issue is already confused about the meaning and function of the `[F]` button on osc. The `(*)` Similarly, there will always be people questioning its meaning. 
However, I confirm that there are scenarios where `forced` and `default` or `external` and `default` coexist, so omitting a few characters makes sense. Therefore, in the end, I chose to keep the `(*)`.

**Comment by tsl0922** [comment]
I pushed 2 commit with your change and some improvements.

**Comment by tsl0922** [comment]
I'll change the default limit value to`0`.
